### PR TITLE
Add support for unity builds

### PR DESCRIPTION
**Pull request by Sharlock93**
**What does this PR do?**

Adds Support for Unity builds in visual studio 2019+. Adds one new flag "enableUnityBuild" with options On and Off.

**How does this PR change Premake's behavior?**

Adds Unity Build/Jumbo Build support.

**Anything else we should know?**

Unity builds really speed up compile time, being able to set them per config and project is a nice feature and something I personally needed.

**Did you check all the boxes?**
Honestly I did this in a hurry and was a neat feature that I wanted myself, if there is anything in particular I should be adding then feedback is welcome.

### CODE REVIEW

**Comment by Jarod42** [comment]
Documentation and unit test are missing.

Wonder if it should be integrated to https://github.com/dcourtois/premake-compilationunit which does [Single compilation unit](https://en.wikipedia.org/wiki/Single_Compilation_Unit) support.

File: src/_premake_init.lua [hunk 1]
```diff
@@ -1512,6 +1512,17 @@
 		tokens = true,
 	}
 
+	api.register {
+		name = "enableUnityBuild",
+		scope = { "config", "rule" },
```
Comment (Jarod42) [inline src/_premake_init.lua:1517]: `"rule"` seems suspicious...

File: src/_premake_init.lua [hunk 2]
```diff
@@ -1512,6 +1512,17 @@
 		tokens = true,
 	}
 
+	api.register {
```
[the remark this answers is not in the document]
Comment (Sharlock93) [reply src/_premake_init.lua]: I honestly didn't see that. Thanks for pointing it out.

File: src/_premake_init.lua (hunk 1 above)
Comment (Sharlock93) [reply src/_premake_init.lua]: Initially when I added this, "config" didn't work, so I added "rule" and that seemed to work. didn't find enough documentation around what it really means but also didn't look that hard.

**Comment by Sharlock93** [comment]
> Documentation and unit test are missing.
> 
> Wonder if it should be integrated to https://github.com/dcourtois/premake-compilationunit (listed in https://premake.github.io/community/modules/#tool-modules) which does [Single compilation unit](https://en.wikipedia.org/wiki/Single_Compilation_Unit) support.

The difference here being that Visual Studio itself does that kind of "Unity/Single Compilation Unit" automatically. For example if you enable it in a project it will generate a file called "unity_<random_suffix>.cpp" that basically does what that module is doing.

I haven't added more flags (honestly because I didn't need them myself) but a more complete set of flags exist for Unity/Jumbo builds that specify how many files should be packed up in one Unity file etc.

image

That module is more useful in a context where native support for Unity builds doesn't exist.

**Comment by Jarod42** [comment]
> > Documentation and unit test are missing.
> 
> If you could point me to where I should be adding those it would be great.

I would say somewhere in premake-core\modules\vstudio\tests\vc2010
 
> > Wonder if it should be integrated to https://github.com/dcourtois/premake-compilationunit (listed in https://premake.github.io/community/modules/#tool-modules) which does [Single compilation unit](https://en.wikipedia.org/wiki/Single_Compilation_Unit) support.
> 
> The difference here being that Visual Studio itself does that kind of "Unity/Single Compilation Unit" automatically. For example if you enable it in a project it will generate a file called "unity_<random_suffix>.cpp" that basically does what that module is doing.
> 
> That module is more useful in a context where native support for Unity builds doesn't exist.

I meant: use special flag for msvc which can do it natively, but keep the module as is for other actions.

**Comment by Sharlock93** [comment]
> I meant: use special flag for msvc which can do it natively, but keep the module as is for other actions.

I mean you could do that but this feature really should be part of visual studio as this is a native feature on visual studio itself so in essence premake is missing it.

**Comment by nickclark2016** [comment]
Can you add the documentation so the CI can pass that step?

File: website/docs/unitybuilds.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
+Enables Unity Builds in Visual Studio, also known as Jumbo Builds
+
+```lua
+enableUnityBuilds "value"
+```
+
+If no toolset is specified for a configuration, the system or IDE default will be used.
```
Comment (Jarod42) [inline website/docs/unitybuilds.md:7]: "toolset"? Seems to be a copy paste...

File: modules/vstudio/vs2010_vcxproj.lua [hunk 4]
```diff
@@ -2550,6 +2551,12 @@
 		end
 	end
 
+	function m.enableJumboBuild(cfg)
+		if _ACTION > "vs2017" and cfg.enableUnityBuild then
```
Comment (Jarod42) [inline modules/vstudio/vs2010_vcxproj.lua:2555]:
```suggestion
		if _ACTION >`= "vs2019" and cfg.enableUnityBuild then
```
To be coherent with surrounding conditions.

File: website/docs/unitybuilds.md (hunk 3 above)
Comment (Sharlock93) [reply website/docs/unitybuilds.md]: yeah 100%, I wanted to keep the same structure and easiest was to copy paste something, been so busy lately that this ended up being a rushed job.

File: modules/vstudio/vs2010_vcxproj.lua (hunk 4 above)
Comment (Sharlock93) [reply modules/vstudio/vs2010_vcxproj.lua:2555]: Ah yes my bad, unity builds exist on 2017, updated the docs.

Comment (Jarod42) [reply modules/vstudio/vs2010_vcxproj.lua:2555]: So 
```suggestion
		if _ACTION >= "vs2017" and cfg.enableUnityBuild then
```

**Comment by nickclark2016** [comment]
Missing one more step in the documentation (and maybe this is something we should better document in the contributing doc), but you need to add it to the `sidebar.js` since it's a new API.

**Comment by Sharlock93** [comment]
> Missing one more step in the documentation (and maybe this is something we should better document in the contributing doc), but you need to add it to the `sidebar.js` since it's a new API.

yeah sorry about that, been super busy with work and it could  be that its written somewhere but I didn't see it.

File: website/docs/Project-API.md [hunk 5]
```diff
@@ -72,6 +72,7 @@
 | [editandcontinue](editandcontinue.md)                     |  |
 | [editorintegration](editorintegration.md)                 | Enable or disable IDE integration |
 | [enablewarnings](enablewarnings.md)                       |  |
+| [enableunitybuilds](unitybuilds.md)                       |  |
```
Comment (KyrietS) [inline website/docs/Project-API.md:75]: Our current convention in docs is that file names are the same as the function that they describe. You have added the `enableUnityBuilds` API, so the file should be named `enableUnityBuilds.md`. Lua is case-sensitive, so the above line is also misleading 😉 Please change it to:

```markdown
| [enableUnityBuilds](enableUnityBuilds.md)                       |  |
```

**Review by KyrietS — CHANGES_REQUESTED** [review]

File: website/docs/Project-API.md (hunk 5 above)
Comment (Sharlock93) [reply website/docs/Project-API.md]: I fixed up all the naming to match what is already there, I hope this is more consistent now.

File: website/sidebars.js [hunk 6]
```diff
@@ -275,6 +275,7 @@ module.exports = {
 						'toolsversion',
 						'transition',
 						'undefines',
+                        'unitybuilds',
```
Comment (KyrietS) [inline website/sidebars.js:278]: `enableunitybuild`

**Review by KyrietS — CHANGES_REQUESTED** [review]
Docs

File: website/sidebars.js (hunk 6 above)
Comment (Sharlock93) [reply website/sidebars.js]: one of these days...

**Review by samsinsane — APPROVED** [review]

**Review by KyrietS — APPROVED** [review]

**Comment by sparkskapil** [comment]
When I try to use this option I get this error.
```attempt to call a nil value (global 'enableunitybuild')```

Am I missing something here?

**Comment by nickclark2016** [comment]
What version of premake are you using?

**Comment by sparkskapil** [comment]
> What version of premake are you using?

Premake 5.0.0-beta2

**Comment by nickclark2016** [comment]
Try one of the CI builds. They should have it.

**Pull request merged**